add a document for machine api usage telemetry

elmiko opened the pull request:
This change brings in an enhancement describing how the Machine API
components can be instrumented to provide usage and behavior information
from telemetry data.

elmiko commented:
/assign @spadgett

elmiko commented:
updates to the user stories, and language around alerting and metrics.

michaelgugino commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,244 @@
+To fully accomplish the goals proposed in this enhancement several metrics
+will need to be exposed through telemetry. These metrics are:
+
+* Machine resource count
```
This IMO is redundant to nodes.

enxebre replied:
For UPI environments automated machine management via machine API is a choice. I think the relation between number of machines and node could give us meaningful insight about the machine API usage in such environments.
I think this could also be valuable in a cluster created via IPI to understand if user cares at all about failed machines. e.g ~100 machines but only ~10nodes. And so identify environments where MHC would come in handy.

smarterclayton replied:
Yeah, it is not redundant to nodes.  Machines and nodes are not the same.

elmiko commented:
updated to remove some language and add more implementation details.

elmiko commented:
putting WIP back on this, i'd like to cleanup the implementation details section a little bit more.

enxebre commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,338 @@
+memory, but haven’t hit max node count) is useful to collect. These events can
+show opportunities to inform users about resource limitations in their clusters.
+
+**Machines with Old Delete Timestamp**
```
fwiw this slightly overlaps with https://github.com/openshift/machine-api-operator/blob/master/install/0000_90_machine-api-operator_04_alertrules.yaml#L23-L32

enxebre commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,338 @@
+
+Epics: Demonstration of Value
+
+Collect data on the total number of scale up and scale down events,
```
to clarify Is this referring to autoscaling scale up/down events, not independent machineSet scale operations?

elmiko replied:
this is about autoscaler events, i'll update the language to be more clear.

JoelSpeed commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,338 @@
+that will aid in building and understanding of how users are operating the
+Machine API components, how frequently they are being exercised, and expose a
+window into the general health of clusters. By reporting the amount of activity
+with the Machine API we can understand the basic use patterns that exist within
```
Would this be clearer as _usage_ patterns?
```suggestion
with the Machine API we can understand the basic usage patterns that exist within
```

JoelSpeed commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,338 @@
+
+### Goals
+
+* Inform about Machine API component(Machine API Operator, Cluster Autoscaler,
```
```suggestion
* Inform about Machine API component (Machine API Operator, Cluster Autoscaler,
```

JoelSpeed commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,338 @@
+or similar event that is impacting many of our users.
+
+For example, since many clusters operate at a relatively small scale, it’s
+possible that losing an AZ in US-East-1 would be small enough event on their
```
Might be good to clarify you mean the AWS region for those who aren't spending every day working with clouds
```suggestion
possible that losing an AZ in AWS US-East-1 would be small enough event on their
```

JoelSpeed commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,338 @@
+experience this simultaneously, we can help provide any potential outage info
+to our users.
+
+**Number of Nodes Covered by the Machine Health Check**
```
How would we know the size of the cluster? Is this kind of data already available?

elmiko replied:
we can calculate the number of nodes from metrics currently exposed, for example [`cluster:node_instance_type_count:sum`](https://github.com/openshift/cluster-monitoring-operator/blob/master/manifests/0000_50_cluster_monitoring_operator_04-config.yaml#L113). there might be other ways to figure this as well, but with the `node_instance_type_count` we could gather all the nodes by disregarding the cardinality of the instance types and just summing all counts.

JoelSpeed commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,338 @@
+
+This metric will help us identify users that are having problems with their
+infrastructure that might go unnoticed. We should trigger an alert if the
+autoscaler cleans up more than X hosts. This alert would be useful for users
```
This will be a rate, so this might make it clearer?
```suggestion
autoscaler cleans up more than X hosts in Y time period. This alert would be useful for users
```

JoelSpeed commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,338 @@
+This is another case where signals across multiple clusters might indicate
+that a cloud or network area is having problems.
+
+For example, Our internal CI infra team had multiple MachineSets configured
+with the autoscaler.  One MachineSet would scale up fine, The other wouldn’t.
+The machines would fail, alarms would fire, the autoscaler would dispose of
+the nodes that failed to come online silently.
```
This paragraph mixes casing on resource names, should keep them consistently camel cased (MachineSet, Machine, Node)

elmiko replied:
i'll clean this up

JoelSpeed commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,338 @@
+
+Epics: Demonstration of Value
+
+Collect data on the total number of scale up and scale down events,
```
Can this be inferred from other data that we already collecting? Eg, are we collecting MachineSet sizing already?

elmiko replied:
i need to update this description but this is specifically about autoscaler events. we can infer scaling by look at the number of Machines tracked by the MAO, but we wouldn't be able to definitely say they were caused by the autoscaler.

JoelSpeed commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,338 @@
+from the Machine API componets. These tests will be simply metric scrapes to
+ensure that the values feeding the telemetry are available.
+
+1. End-to-end tests that ensure the proper alerts are firing under the conditions
```
```suggestion
2. End-to-end tests that ensure the proper alerts are firing under the conditions
```

JoelSpeed commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,338 @@
+their specific conditions. These tests will ensure that our alerting is not
+drifting with respect to the underlying events.
+
+2. Functional tests of created artifacts to prove continued operation. There
```
```suggestion
3. Functional tests of created artifacts to prove continued operation. There
```

elmiko commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,338 @@
+## Summary
+
+This enhancement proposes utilizing alerts and metrics exposed through telemetry
+that will aid in building and understanding of how users are operating the
```
yes

elmiko commented:
> This document refers to us seeing the data and being able to compare across clusters, for that to happen, we need to configure the metrics to be sent back to Red Hat, how is that done? Do we need to update or add a particular resource? Should this be mentioned somewhere?

the next step after we identify the metrics we'd like to expose is to create a pr in the cluster-monitoring-operator repo and add our metrics to [this file](https://github.com/openshift/cluster-monitoring-operator/blob/master/manifests/0000_50_cluster_monitoring_operator_04-config.yaml). the alerts are automatically exported.

i've captured some of the metric values we want to export in the implementation section, i can add an extra sentence there about the next part if it makes sense. i left it out since i think the main details are the metrics we want to expose, the alerts we can use, and what we want to do with them.

elmiko commented:
updated to address comments and clean up some of the language

elmiko commented:
removed more unnecessary language, and cleaned up a few inconsistencies. i am removing WIP as i feel this is ready for more review.

michaelgugino commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,340 @@
+and see if we can help with their configuration. This can also help us identify
+areas where increased education will have value.
+
+**Failed Node Cleanups**
```
It's not clear this story is related to the cluster-autoscaler just from reading.

elmiko replied:
i've updated the first sentence to be more clear, let me know if that works

elmiko commented:
updated the first sentence of the Failed Node Cleanups entry

JoelSpeed commented:
/lgtm

enxebre commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,340 @@
+
+#### Analysis and Publication
+
+Collecting the metrics associated with this enhancment is one part of the
```
typo `enhancment`

enxebre commented:
Thanks @elmiko! this looks good to me as to start to have more detailed discussions over the counterpart PRs.
/approve
/hold
To give others the chance to chime in. @elmiko feel free to un hold.
PTAL @openshift/openshift-team-monitoring @smarterclayton @wking

lilic commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,340 @@
+would dispose of the nodes that failed to come online. These failed nodes were
+being silently cleaned up by the autoscaler.
+
+**Total Scale Events**
```
I am missing in all of these the actual metric that will be used here. Are these new metrics that you will introduce, or existing ones?  


Not sure you seen but we have telemetry docs for each team, how to measure number of series, the amount of series to be sent etc. https://docs.google.com/document/d/1a6n5iBGM2QaIQRg9Lw4-Npj6QY9--Hpx3XYut-BrUSY/edit

elmiko replied:
> I am missing in all of these the actual metric that will be used here. Are these new metrics that you will introduce, or existing ones?

i've tried to break the implementation section into a couple parts, this first part talks about how we will use the exported metrics and what calculations we would like to determine from them. in the following section "Alerts, Metrics, and Telemetry" i start to enumerate the specific metrics we will need to export. i can add some details about specific metric names there, would that work?

lilic replied:
It's up to you, we can also do it in the PRs, happy to review those, or here. There are certain best practices around metric naming. 
>  i can add some details about specific metric names there, would that work?

Its more about are these new metrics or are you using existing metrics to send. So this would be helpful to use actual metric names yes. It's more about to answer questions like:
- could these metrics be done upstream
- maybe we have these metrics already, so we don't duplicate the efforts
- best practices around naming these metrics if they are new 
- are these going to be high unbound cardinality metrics

hence examples of actual metrics and all their possible labels help with both in cluster prometheus ingestion as well as keeping limit of series for those we sending to telemetry.

elmiko replied:
> Its more about are these new metrics or are you using existing metrics to send. So this would be helpful to use actual metric names yes. It's more about to answer questions like:
> 
>     * could these metrics be done upstream
> 
>     * maybe we have these metrics already, so we don't duplicate the efforts
> 
>     * best practices around naming these metrics if they are new
> 
>     * are these going to be high unbound cardinality metrics

thanks, this is good advice.

i think i talk about _some_ of these details below, but i will go back and make sure that it is clear we are answering these questions.

lilic commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,340 @@
+This metric can show us stale resources that exist within clusters. By collecting
+data for machines that are stuck in a deletion phase for prolonged periods of time
+we can identify opportunities to inform users and create alerts around these
+stale resources. This alerting might need to be tunable (if possible), but a
```
Alerts cannot be tunable. You can have different levels of severity. So for example per threshold you can have severity info, higher threshold can have severity critical which would mean someone needs to be paged.

elmiko replied:
ack, i can remove the language about tunable alerts.

lilic commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,340 @@
+
+Epics: Proactive Support
+
+This metric can show us stale resources that exist within clusters. By collecting
```
The proposed alert feels like alerting on cause already. I would prefer alerting on the symptom of why we get stale resources so we prevent them from happening. See more about cause vs symptom based alerting here -> https://prometheus.io/docs/practices/alerting/

elmiko replied:
that makes sense to me. this specific use case seems like a chicken/egg problem at some level. iirc, this metric story is about what we observe in the field (eg at some times clusters have these machines caught in stale deletion states) and adding some analysis to the failures that happens. we know that there is a root cause _somewhere_ but do not have an understanding of how it happened yet, but we know we have machine that get caught in this state. i _think_ that was what this metric was driving towards.

perhaps @michaelgugino could add some extra context here

michaelgugino replied:
I disagree this would be alerting for 'cause' instead of 'symptom.'  The 'symptom' is a machine with a delete timestamp older than X.  The cause could be any number of reasons such as expired cloud credentials, undrainable node, machine-controller died, etc.

wking replied:
https://prometheus.io/docs/practices/alerting/ links [this doc][1], which has a section on *Alerting from the spout* that describes cause vs. symptom applying to external users consuming cluster-provided services.  Stale machines seem far from the spout.  It make sense to me as a warn-level alert, just not a wake-me-at-midnight critical alert.  I don't see anything in this enhancement today that specifically calls out new alerts or their intended levels.

[1]: https://docs.google.com/a/boxever.com/document/d/199PqyG3UsyXlwieHaqbGiWVa8eMWi8zzAn0YfcApr8Q

elmiko replied:
>  I don't see anything in this enhancement today that specifically calls out new alerts or their intended levels.

i'm still gathering the details for the alerts, hoping to add them today.

michaelgugino replied:
I agree that for the case of 1 or 2 machines with an old delete timestamp, that could probably be a warning.  If the cluster-autoscaler or MHC are short-circuited because of excessive machines/nodes being unhealthy, that would be a critical.

If a control-plane machine is in anyway not healthy, that would be a critical.  We definitely need to iron out some of the details in some of our existing alerts and some of the new ones.

wking replied:
> If the cluster-autoscaler or MHC are short-circuited because of excessive machines/nodes being unhealthy, that would be a critical.

So make a critical alert for "there is no schedulable space left"?  Maybe the cluster is still serving clients happily even without more space, and mitigation can wait another hour?

> If a control-plane machine is in anyway not healthy, that would be a critical.

This is one hiccup away from quorum loss, so agreed here.  But maybe etcd should be critical about quorum-vulnerable and CVO would be critical about "failing to schedule operator deployment" and the MHC warning would just inform recovery from those spout-ier critical issues?

smarterclayton replied:
General statement to lili's comment (going even broader):

In a declarative system, whenever the declarative goal cannot be realized within time X, that is an alertable condition.  It may not be a great alert in all cases, but alerted >>> unalerted.  We can always improve the alert later.  However, the key criteria is "does not converge" is ALWAYS better to know about than not, because declarative systems depend on convergence to satisfy user intent.

I would like us to make sure we talk about this more broadly, but this is probably as important as "USE" to kubernetes alerting and metrics - "failure to converge" is the most important error in "E", and all teams should ask "do you alert when you do not converge within a reasonable amount of time".  I struggle to think of more than a handful of examples where failure to converge is non-alertable.

brancz commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,340 @@
+* MachineHealthCheck resource count
+* MachineHealthCheck remediation success count
+* MachineHealthCheck remediation failure count
+* Cluster Autoscaler scale event count (with labels for different event types)
```
could you elaborate on the number of labels, as that will affect the total cardinality

elmiko replied:
yes, i was imagining only 2 options for this label. it would be something like `cluster_autoscaler_event_count{type="out"}` and `cluster_autoscaler_event_count{type="in"}`. those are just example names.

should i add this detail to the doc?

brancz replied:
yes please, that sounds fine, it's just to get an accurate picture of the maximum number of series we would be sending

brancz commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,340 @@
+* MachineSet resource count
+* MachineAutoscaler resource count
+* MachineHealthCheck resource count
+* MachineHealthCheck remediation success count
```
these should not be sent separately, but instead via an alert when the failure rate is too high to be acceptable

elmiko replied:
that makes sense to me, and i was planning that we would use the alerts to augment this data.

would it be acceptable to return total remediations? (i'm just thinking about how do we calculate the remediation effectiveness without know both total attempts and failures)

brancz replied:
yes I think that would be ok, try to aim for a maximum of 10 series aside from alerts

elmiko replied:
i am working on adding more information about the specific metrics and series we would like to generate from this enhancement. hopefully that will help add more context to the specifics (cc @lilic )

elmiko commented:
just an update, i am continuing to collect some information about specific metric and alert names. hoping to have an update tomorrow.

elmiko commented:
updated to add more specific metrics names and series in the implementation details section. also removed the language about configurable alerts in the specific telemtry inquiries section.

i am still investigating the alert names that exist and the areas where we will need to add more alerting. hopefully i can add that tomorrow, but i think the bulk of thie enhancement is in place.

michaelgugino commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,365 @@
+users know we’re seeing this across many clusters would assist in their own troubleshooting
+efforts and save them time.
+
+**Remediation is Disabled or Short Circuited**
```
I didn't see this metric/alert called out in the list of things to be added.

elmiko replied:
this is going to be gathered from an alert, i'm still working on those.

brancz commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,365 @@
+* Total MachineAutoscaler resource count, using `cao_machineautoscaler_items` with no labels.
+* Total MachineHealthCheck resource count, using `mapi_machinehealthcheck_items` with no labels.
+* Total nodes covered by MachineHealthChecks count, using `mapi_machinehealthcheck_total_nodes_count` with no labels.
+* Total remediations completed by MachineHealthChecks count, using `mapi_machinehealthcheck_remediation_success_count` with no labels.
```
`_count` is a reserved suffix, best practice is not to use it to not confuse users, you just want something along the lines of `mapi_machinehealthcheck_remediation_successes` (unless this is a true counter, in which case this should be the `_total` suffix instead of `_count`)

elmiko replied:
ack, i will update these. thanks!

brancz replied:
I should have also been clearer in my first comment here, counters are basically always better if you can model your data as that, and this seems like a perfect use case.

elmiko commented:
updated to include the alert information and fix a few naming issues with the metrics.

elmiko commented:
fixed typo and a couple missed metric names

elmiko commented:
i am removing the hold on this as our team has discussed internally and we feel this ready to progress forward.
/hold cancel

elmiko commented:
updated the header block to reflect the reviewers here, thanks everyone!

elmiko commented:
@JoelSpeed @michaelgugino @lilic @brancz @wking 

just curious if the latest changes i've made have answered the open questions, thanks again for the reviews =)

lilic commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,390 @@
+**These alerts exist today**
+
+* machine-api-operator
+    * MachineWithooutValidNode
```
MachineWithoutValidNode?
```suggestion
    * MachineWithooutValidNode
```

JoelSpeed commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,390 @@
+
+### Non-Goals
+
+* Provide any information that would expose a user's identify or confidential
```
Typo?
```suggestion
* Provide any information that would expose a user's identity or confidential
```

elmiko commented:
updated with typo fixes

> FYI usually Clayton gives final +1 on all things sent via telemetry, as per the docs I shared.

thanks for the reminder @lilic!

@smarterclayton ptal

smarterclayton commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,390 @@
+**These metrics exist today**
+
+* Machine resource count
+    * `mapi_machine_items` - This metric has no labels.
```
You should not do this, you should use etcd_resource_count and map it into the exsiting cluster_resource_usage metric that tracks specific resources.  Don't duplicate existing core metrics, but you should add recording rules that map them.

See cluster:usage:resources:sum where we whitelist specifc resources.  MHC and machines should be in that list.

elmiko replied:
ack, i will need to read up on etcd_resource_count, but i will adjust this enhancement to reflect that.

smarterclayton commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,390 @@
+**These metrics will need to be created**
+
+* MachineAutoscaler resource count
+    * `cao_machineautoscaler_items` - This metric has no labels.
```
See previous comment about cluster:usage:resources:sum

smarterclayton commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,390 @@
+**Metric series to be exported**
+
+* Total Machine resource count, using `mapi_machine_items` with no labels.
+* Total MachineSet resource count, using `mapi_machineset_items` with no labels.
```
See previous comment about cluster:usage:resources:sum

smarterclayton commented on enhancements/machine-api/machine-api-usage-telemetry.md:
```diff
@@ -0,0 +1,390 @@
+  as a single combined metric of all labels.
+* Total Cluster Autoscaler scale up nodes, using `cluster_autoscaler_scaled_up_nodes_total`
+  with no labels.
+* Total MachineAutoscaler resource count, using `cao_machineautoscaler_items` with no labels.
```
See previous comment about cluster:usage:resources:sum

smarterclayton commented:
This is a great proposal, a few comments inline, but this is approved for adding this cardinality (counts and such need to fit into the existing metrics).

elmiko commented:
i've updated the enhancement to reference the time series `cluster:usage:resources:sum` for the various CRD counts we want.

elmiko commented:
@smarterclayton @lilic @brancz is this in good shape to merge now? =)

s-urbaniak commented:
/lgtm
on behalf of the requested reviewers (lili is out sick) as the comments seem to have been addressed and clayton gave its ok.

s-urbaniak commented:
(in case we have still concerns we'd still do a follow-up)

s-urbaniak commented:
/lgtm
